### PR TITLE
Story 3

### PR DESCRIPTION
**Pull request by shaojunda**
implement upload.

### CODE REVIEW

**Comment by shaojunda** [comment]
implement bookmark list

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+         :recoverable, :rememberable, :trackable, :validatable
+
+  has_many :product_relations
+  has_many :bookmarked_products, :through => :product_relations, :source => :product
+
+  def add_bookmark!(product)
+    bookmarked_products << product
+  end
+
+  def removebookmark!(product)
```
Comment (xdite) [inline app/models/user.rb]: 這裡最好寫成 remove_bookmark!

File: app/controllers/admin/uses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  def to_admin
+    @user = User.find(params[:id])
+    @user.to_admin!
+    redirect_to admin_uses_path, notice: "updated to admin success"
+  end
+
+  def to_user
+    @user = User.find(params[:id])
+    @user.to_user!
```
Comment (xdite) [inline app/controllers/admin/uses_controller.rb]: degrade_to_user!

File: app/controllers/admin/uses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  before_action :require_admin
+  layout "admin"
+
+  def index
+    @users = User.all
+  end
+
+  def to_admin
+    @user = User.find(params[:id])
+    @user.to_admin!
```
Comment (xdite) [inline app/controllers/admin/uses_controller.rb]: upgrade_to_admin!

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+
+  def bookmarks
+    @products = current_user.bookmarked_products
+  end
+
+  def like
+    @product = Product.find(params[:id])
+
+    if !current_user.is_bookmark_of?(@product)
```
Comment (xdite) [inline app/controllers/users_controller.rb]: alrady_bookmarked?(@product)

File: app/models/user.rb (hunk 1 above)
Comment (shaojunda) [inline app/models/user.rb]: 已修改。

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  before_action :authenticate_user!
+
+
+  def bookmarks
+    @products = current_user.bookmarked_products
+  end
+
+  def like
+    @product = Product.find(params[:id])
+
```
Comment (xdite) [inline app/controllers/users_controller.rb:11]: FavoritesController  CD

File: app/models/cart.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    ci.save
+  end
+
+  def del_cart_items(ids)
+    # puts "+++++#{ids}"
+    # cart_items = CartItem.find_by_sql("select * from cart_items where cart_id = #{self.id} and id in (#{ids})").delete_all
+    CartItem.find_by_sql("delete from cart_items where cart_id = #{self.id} and id in (#{ids})")
+    # cart_items.find_by(id: "#{ids}")
+  end
+
```
Comment (xdite) [inline app/models/cart.rb:39]: cart_items.delete_all
CartItem.wehre(:id => ids )

File: app/models/cart_item.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#  product_id :integer
+#  cart_id    :integer
+#
+
+class CartItem < ApplicationRecord
+  belongs_to :product
+  validates :quantity, presence: true, numericality: {greater_than: 0}
+
+
+  def up_count!
```
Comment (xdite) [inline app/models/cart_item.rb:18]: increase_amount!

File: app/models/cart_item.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      self.quantity -= 1
+      self.save
+    end
+  end
+
+  def edit_cart_item_count!(quantity)
+    product = self.product.quantity
+    if !product.blank? && !quantity.blank? && quantity.to_i > 0 && quantity.to_i < product
+      self.quantity = quantity
+      if self.save()
```
Comment (xdite) [inline app/models/cart_item.rb:37]: if save

Comment (xdite) [inline app/models/cart_item.rb:37]: if   some condition 
 return xxx
end

if save
else
end

File: app/models/product.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  validates :price, numericality: true
+
+  has_many :product_relations
+  has_many :members, through: :product_relations, source: :user
+
+  scope :is_publish, -> {where is_publish: true}
+
+  def publish!
+    if !self.is_publish
+      self.is_publish = true
```
Comment (xdite) [inline app/models/product.rb:33]: update_column(:is_publish, true )

**Pull request closed without merging**